lib/vsc/filesystem/oceanstor.py

```diff
@@ -151,11 +172,229 @@ def __init__(self, url, username=None, password=None):
+        self.log.debug("Filtering filesystems in storage pools with IDs: %s", ', '.join(str(i) for i in sp_ids))
+
+        if not update and self.filesystems:
+            # Use cached filesystem data
+            filesystems = {fs['name']: fs for fs in self.filesystems.values() if fs['storage_pool_id'] in sp_ids}
+            dbg_prefix = "(cached) "
+        else:
+            # Request filesystem data
+            _, response = self.session.file_service.file_systems.get(filter=filter_sp_ids_json)
+            filesystems = {fs['name']: fs for fs in response['data']}
+            dbg_prefix = ""
+
+            self.filesystems = filesystems
+
+        self.log.debug(dbg_prefix + "Filesystems in OceanStor: %s", ", ".join(filesystems))
+
+        return filesystems
+
+    def select_filesystems(self, filesystemnames, devices=None, byid=False):
+        """
+        Return list of existing filesytem with the provided filesystem names
+        Restrict list to filesystems found in given storage pools names
+
+        @type filesystemnames: list of filesystem names (if string: 1 filesystem; if None: all known filesystems)
+        @type devices: list of storage pools names (if string: 1 storage pool; if None: all known storage pools)
+        @type byid: boolean (if True: return list of filesystem IDs)
+        """
+
+        if not isinstance(filesystemnames, list):
+            target_filesystems = [filesystemnames]
+        else:
+            target_filesystems = filesystemnames
+
+        # Filter by storage pools
+        filesystems = self.list_filesystems(device=devices)
+
+        try:
+            fs_select = [filesystems[fs]['name'] for fs in target_filesystems]
+        except KeyError as err:
+            fs_miss = err.args[0]
+            fs_avail = ", ".join(filesystems)
+            errmsg = "Filesystem '%s' not found in OceanStor. Use any of: %s" % (fs_miss, fs_avail)
+            self.log.raiseException(errmsg, KeyError)
+
+        # Convert names to IDs
+        if byid:
+            fs_select = [filesystems[fs]['id'] for fs in fs_select]
+
+        return fs_select
+
+    def list_filesets(self, devices=None, filesystemnames=None, filesetnames=None, update=False):
         """
-        List all filesystems.
+        Get all dtree filesets in given devices and given filesystems
+        Filter reported results by name of filesystem
+        Store unfiltered data in self.fileset (all filesets in given devices and given filesystems)
+
+        @type devices: list of devices (if string: 1 device; if None: all found devices)
+        @type filesystemnames: list of filesystem names (if string: 1 filesystem; if None: all known filesystems)
+        @type filesetnames: list of fileset names (if string: 1 fileset)
+
+        Set self.filesets as dict with
+        : keys per parent filesystemName and value is dict with
+        :: keys per dtree fileset ID and value is dict with
+        ::: keys returned by OceanStor:
```

Comment:
shouldn't 'path' also be in here?

Reply:
That's a good question. It is possible to define the path of the parent directory holding a dtree on creation (`parent_dir` option). However, the value of `parent_dir` is not returned by the API when querying for dtrees (`file_service/dtrees`). So, at this stage, it is impossible for us to know if a dtree is mounted in the root of its filesystem or not.